src/v1/types/requests.rs

```
@@ -0,0 +1,119 @@
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(rename_all = "SCREAMING_SNAKE_CASE"))]
+pub enum View {
+    /// The `MINIMAL` view.
+    #[default]
+    Minimal,
+
+    /// The `BASIC` view.
+    Basic,
+
+    /// The `FULL` view.
+    Full,
+}
+
+/// A method to get the default view of parameters.
+const fn default_view() -> View {
+    View::Minimal
+}
+
+/// The query parameters for `GetTask` endpoint.
+#[derive(Copy, Clone, Debug, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct GetTaskParams {
+    /// The view of the returned task.
+    #[cfg_attr(feature = "serde", serde(default = "default_view"))]
+    pub view: View,
+}
+
+/// The query parameters for `ListTasks` endpoint.
+#[derive(Clone, Debug, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ListTasksParams {
+    /// The filter for task name (prefixed).
+    ///
+    /// If unspecified, no task name filtering is done.
+    pub name_prefix: Option<String>,
+    /// The filter for task state.
+    ///
+    /// If unspecified, no task state filtering is done.
+    pub state: Option<State>,
+    /// The filter for task tag keys.
+    ///
+    /// This is zipped with `tag_values`.
+    ///
+    /// If empty, no task tags filtering is done.
+    #[cfg_attr(feature = "serde", serde(rename = "tag_key", default))]
+    pub tag_keys: Vec<String>,
+    /// The filter for task tag values.
+    ///
+    /// This is zipped with `tag_keys`.
+    ///
+    /// If the value is empty, it matches all values.
+    ///
+    /// It is an error if more values are supplied than keys.
+    #[cfg_attr(feature = "serde", serde(rename = "tag_value", default))]
+    pub tag_values: Vec<String>,
+    /// The number of tasks to return in one page.
+    ///
+    /// Must be less than 2048. Defaults to 256.
```

What happens when this is greater than 2048? Is some sort of validation issue thrown?

---

I came back around to this and don't recall seeing anything. It would be good to throw a validation error potentially and add a test for it, but I wouldn't say it's a hard requirement for the PR to go through.

---

The service should likely fail with a 400, although errors aren't really specified in the spec.

---

I agree, but I'm also thinking that, since it is specified in the specification that this muss be less than 2048, we could throw a deserialization/validation error if an invalid value is tried at deserialization/creation. Is there a good reason _not_ to do that?

---

I think that perhaps it'd be best to bake that into `list_tasks` in client rather than serialization / deserialization of `ListTaskParams`.

If we bake it into deserialization, for example, depending on the server framework, the server might not have the best control over the response returned.

---

I pushed up a validation in the client.

---

Okay, I can live with that.